tools/tpm2_createpolicy.c

```diff
@@ -0,0 +1,333 @@
+#include <ctype.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "log.h"
+#include "options.h"
+#include "main.h"
+#include "string-bytes.h"
+#include "tpm_session.h"
+#include "tpm_hash.h"
+#include "files.h"
+
+
+typedef struct createpolicypcr_ctx createpolicypcr_ctx;
+struct createpolicypcr_ctx {
+    //common policy options
+    TSS2_SYS_CONTEXT *sapi_context;
+    SESSION *policy_session;
+    TPM2B_DIGEST policy_digest;
+    char policyfile[PATH_MAX];
+    bool policy_file_flag;
+    //build_pcr_policy options
+    unsigned int pcr_index;
+    TPMI_ALG_HASH hash_alg;
+    char raw_pcr_file[PATH_MAX];
+    bool policy_type_pcr_flag;
+    bool raw_pcr_flag;
+    bool pcr_index_flag;
+};
+
+#define INIT_SIMPLE_TPM2B_SIZE( type ) (type).t.size = sizeof( type ) - 2;
+
+#define SET_PCR_SELECT_BIT( pcrSelection, pcr ) \
+ (pcrSelection).pcrSelect[( (pcr)/8 )] |= ( 1 << ( (pcr) % 8) );
+
+TPM_RC build_pcr_policy( createpolicypcr_ctx *ctx) {
+
+    TPML_PCR_SELECTION pcrs = {
+        .count = 1,
+        .pcrSelections[0].hash = ctx->hash_alg,
+        .pcrSelections[0].sizeofSelect = 3, 
+        .pcrSelections[0].pcrSelect[0] = 0, 
+        .pcrSelections[0].pcrSelect[1] = 0, 
+        .pcrSelections[0].pcrSelect[2] = 0 
+    };
+    SET_PCR_SELECT_BIT( pcrs.pcrSelections[0], ctx->pcr_index);
+
+    //digests[count{2:}]{:8}
+    TPML_DIGEST pcrValues = {
+        .count = pcrs.count
+    };
+    UINT32 pcrUpdateCounter;
+    TPM2B_DIGEST pcrDigest;
+    TPML_PCR_SELECTION pcrSelectionOut;
+    
+    TPM_RC rval = TPM_RC_SUCCESS;
+    FILE *fp;
+    unsigned i=0, sz=0;
+    long filesize=0;
+    if(ctx->raw_pcr_flag) {
+        rval = getFileSize(ctx->raw_pcr_file, &filesize);
+        if(rval!=TPM_RC_SUCCESS) {
+            return rval;
+        }
+
+        switch(ctx->hash_alg) {
+            case TPM_ALG_SHA1: 
+                pcrValues.digests[0].t.size = SHA1_DIGEST_SIZE;
+                break;
+            case TPM_ALG_SHA256:
+                pcrValues.digests[0].t.size = SHA256_DIGEST_SIZE;
+                break;
+            case TPM_ALG_SHA384:
+                pcrValues.digests[0].t.size = SHA384_DIGEST_SIZE;
+                break;            
+            case TPM_ALG_SHA512:
+                pcrValues.digests[0].t.size = SHA512_DIGEST_SIZE;
+                break;
+            case TPM_ALG_SM3_256:
+                pcrValues.digests[0].t.size = SM3_256_DIGEST_SIZE;
+                break;
+            default:
+                pcrValues.digests[0].t.size = SHA1_DIGEST_SIZE;
+                break;
+        }
+
+        if(filesize!=pcrValues.digests[0].t.size) {
+            LOG_ERR("Input PCR file %s size mismatched the chosen PCR algorithm\n",
+                ctx->raw_pcr_file);
+            return -1;
+        }
+
+        fp = fopen(ctx->raw_pcr_file, "rb");
+        for(i=0; i<pcrValues.digests[0].t.size; i++) {
+            sz = fread(&pcrValues.digests[0].t.buffer[i],1,1,fp);
```

Comment:
byte for byte reading... with no check on fread return.

size_t sz = fread(&pcrValues.digests[0].t.buffer[i], 1, pcrValues.digests[0].t.size);
if (sz !=  pcrValues.digests[0].t.size) {
   if (is_ferror(fp) {
   LOG_ERR("log errno here");
   } else {
    LOG_ERR("File size not expected, got: %zu expected: %u", sz, pcrValues.digests[0].t.size);
   goto out; //At label out close the open file... or fclose it on these error cases...
   }
}